test/integration/bad/functions-bad18.stan

```diff
@@ -4,9 +4,9 @@ functions {
   }
 }
 parameters {
-  real foo;  // should be name conflict here
+  real foo;
 }
 model {
   foo ~ normal(0.0, 1.0);
-  foo();
+  foo();  // the function has been shadowed, this `foo` is a number
```

Comment:
That math library "contextual name shadowing" doesn't make much sense to me and I couldn't find if it's documented anywhere.

Reply:
I'm not sure why it was done this way, but my understanding is:

1. It was desirable to let users name variables the same as built in functions, because names like `beta` are common
2. The fact that C++ namespaces allow us to differentiate between a function (in `stan::math`) and a variable (in `::`) meant we could also let users call that function too. So, we allow it. 

Essentially, [Stan is a Lisp2](https://www.nhplace.com/kent/Papers/Technical-Issues.html), at least when it comes to the standard library. Emulating this same behavior in user functions is possible, but would require extra work due to the fact that we don't namespace-qualify UDF calls.